apm-sniffer/apm-sdk-plugin/thrift-plugin/src/main/java/org/apache/skywalking/apm/plugin/thrift/wrapper/ServerInProtocolWrapper.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.thrift.wrapper;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.StringTag;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TField;
+import org.apache.thrift.protocol.TMap;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TType;
+
+/**
+ * Wrapping server input protocol for reading and parsing the trace header from the original protocol. It is a special
+ * field which will be skipped without one deal with.
+ */
+public class ServerInProtocolWrapper extends AbstractProtocolWrapper {
+    private static final ILog LOGGER = LogManager.getLogger(ServerInProtocolWrapper.class);
+    private static final StringTag TAG_ARGS = new StringTag("args");
+    private AbstractContext context;
+
+    public ServerInProtocolWrapper(final TProtocol protocol) {
+        super(protocol);
+    }
+
+    public void initial(AbstractContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public TField readFieldBegin() throws TException {
+        final TField field = super.readFieldBegin();
+        if (field.id == SW_MAGIC_FIELD_ID && field.type == TType.MAP) {
+            try {
+                TMap tMap = super.readMapBegin();
+                Map<String, String> header = new HashMap<>(tMap.size);
+
+                for (int i = 0; i < tMap.size; i++) {
+                    header.put(readString(), readString());
+                }
+
+                AbstractSpan span = ContextManager.createEntrySpan(
+                    context.getOperatorName(), createContextCarrier(header));
+                span.start(context.startTime);
+                span.tag(TAG_ARGS, context.getArguments());
+                span.setComponent(ComponentsDefine.THRIFT_SERVER);
+                SpanLayer.asRPCFramework(span);
+            } catch (Throwable throwable) {
+                LOGGER.error("Failed to resolve header or create EntrySpan.", throwable);
```

Comment:
All catches are not necessary in the interceptor scope, as the core has try/catch-log already.

Reply:
These wrapping objects are executed outside of interceptor.